### DISCUSSION TITLE
Startup delay for some applications

### DISCUSSION THREAD

**Discussion by dancaer69** [WayfireWM/wayfire#1807] [Q&A]
Hello, I recently started using wayfire and I have the annoying issue of delayed startup by 20-30 secs. I found an issue(#775) about it and added some variables but I still have this issue. At first the delay happens every time I run a gtk application. After I used dbus-launch-session to start wayfire this solved but I still have problem when login to wayfire. I'm using waybar and eww and waybar needs about 30secs to start at login. Also firefox, nemo(and propably other gtk apps) delay about 30secs if I start them at this time(I mean before waybar appear on screen). Same think happens with wf-background(need about 30secs to start at login), but swaybg starts immediately.
Based on suggestions I have those added when start wayfire:
```
export XDG_SESSION_TYPE=wayland
	export XDG_SESSION_DESKTOP=Wayfire
	export XDG_CURRENT_DESKTOP=Wayfire

	# Wayland stuff
	export MOZ_ENABLE_WAYLAND=1
	export QT_QPA_PLATFORM=wayland
	export SDL_VIDEODRIVER=wayland
	export _JAVA_AWT_WM_NONREPARENTING=1
    dbus-run-session wayfire
```
Also those in wayfire.ini:
```
0_environment = dbus-update-activation-environment --systemd WAYLAND_DISPLAY DISPLAY XAUTHORITY
portal = /usr/lib/xdg-desktop-portal-wlr -r
```
But neither of those solve the problem. The only think which helped was the "dbus-run-session", and it solved the general startup delay of gtk apps only. 
Any help?

**Comment by soreau** [comment]
Have you tried uninstalling xdg-desktop-portal-gnome?

**Reply by dancaer69** [reply]
It's not installed.

**Comment by leafi** [comment]
Yeah, I don't know the right way to solve this, but I have in Autostart an entry for a script in my home directory, `wayfire-startup.sh` (remember to chmod +x it.). My current script looks like this (minus logging stuff):

```
#!/bin/sh

systemctl --user import-environment DISPLAY WAYLAND_DISPLAY
hash dbus-update-activate-environment && dbus-update-activation-environment --systemd DISPLAY WAYLAND_DISPLAY
/usr/lib/xdg-desktop-portal-gtk &
sleep 1

# my script: I then launch gammastep and run_keybase, forked (line ends in &)

/usr/lib/polkit-gnome/polkit-gnome-authentication-agent-1 &
# mako blocks...
mako &
```

That's enough for Nautilus, Firefox, GNOME Disks, things in Flatpaks etc. to work correctly, without long delays.

(BTW, if  you find the Nautilus file browser can't open Archive Manager or the image previewer eog when you double-click on files, that's a bug in a (recent? current?) version of the GNOME stack. I believe it's fixed as of 44.2).

There's lots of log lines & stdout & stderr redirection in my actual script, because determining what forked and what blocked was a trial-and-error process - as well as why things failed when they did fail.

This effectively all runs after Wayfire has already started (but before wf-shell starts?). I start Wayfire with lightdm & lightdm-gtk-greeter which I believe uses the .desktop file.

**Reply by dancaer69** [reply]
I don't use any desktop manager and this you posted didn't make any difference either. I have a ~30 seconds delay on gtk applications(at least nemo, waybar, firefox which I test) on startup. After that there is no delay. This delay happens and if I use wayfire's panel(which I don't but I tested it).

**Comment by ammen99** [comment]
I have set up everything in my autostart section to wait for 1 second after running the dbus update command (because dbus communication takes some time):

```ini
[autostart]
0_0 = systemctl --user import-environment
autostart_wf_shell = false

bg = sleep 1 && wf-background
gtk = sleep 1 && (XDG_SESSION_TYPE=wayland XDG_CURRENT_DESKTOP=sway /usr/libexec/xdg-desktop-portal --replace & /usr/libexec/xdg-desktop-portal-wlr)
mako = sleep 1 && mako
panel = sleep 1 && wf-panel
```

I am not sure whether this helps in your case, but it is worth trying.

**Reply by dancaer69** [reply]
Unfortunately didn't do any difference. The problem isn't with the autostart applications only. If I manually start nemo or firefox, they don't start before pass those 30 seconds.
